From 3f2af23a85a931de429ddf632cbd64c470401d8b Mon Sep 17 00:00:00 2001
From: Quintus Chu <ardanzhu@gmail.com>
Date: Tue, 29 Dec 2020 05:45:01 +0800
Subject: [PATCH] 	shortcut-fe 
 target/linux/generic/hack-5.4/952-net-conntrack-events-support-multiple-registrant.patch
 
 target/linux/generic/hack-5.4/953-net-patch-linux-kernel-to-support-shortcut-fe.patch

---
 .../base-files/files/etc/init.d/shortcut-fe   | 52 +++++++++++++++++++
 1 file changed, 52 insertions(+)
 create mode 100644 package/base-files/files/etc/init.d/shortcut-fe

diff --git a/package/base-files/files/etc/init.d/shortcut-fe b/package/base-files/files/etc/init.d/shortcut-fe
new file mode 100644
index 0000000000..83f628bc58
--- /dev/null
+++ b/package/base-files/files/etc/init.d/shortcut-fe
@@ -0,0 +1,52 @@
+#!/bin/sh /etc/rc.common
+#
+# Copyright (c) 2014-2015 The Linux Foundation. All rights reserved.
+# Permission to use, copy, modify, and/or distribute this software for
+# any purpose with or without fee is hereby granted, provided that the
+# above copyright notice and this permission notice appear in all copies.
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
+# OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+#
+
+#SFE connection manager has a lower priority, it should be started after other connection manager
+#to detect the existence of connection manager with higher priority
+START=90
+
+have_cm() {
+	[ -d "/sys/kernel/debug/ecm" ] && echo 1 && return
+
+	echo 0
+}
+
+#load shortcut-fe connection manager
+load_sfe_cm() {
+	local kernel_version=$(uname -r)
+
+	#shortcut-fe-drv.ko is not needed because other connection manager is not enabled
+	[ -d "/sys/module/shortcut_fe_drv" ] && rmmod shortcut_fe_drv
+
+	[ -e "/lib/modules/$kernel_version/fast-classifier.ko" ] && {
+		[ -d /sys/module/fast_classifier ] || insmod /lib/modules/$kernel_version/fast-classifier.ko && return
+	}
+
+	[ -e "/lib/modules/$kernel_version/shortcut-fe-cm.ko" ] && {
+		[ -d /sys/module/shortcut_fe_cm ] || insmod /lib/modules/$kernel_version/shortcut-fe-cm.ko && return
+	}
+}
+
+start() {
+	[ "$(have_cm)" = "1" ] || load_sfe_cm
+	echo 1 > /sys/fast_classifier/skip_to_bridge_ingress
+	sfe_ipv6=$(cat /sys/sfe_ipv6/debug_dev)
+	[ ! -f /dev/sfe_ipv6 ] && mknod /dev/sfe_ipv6 c $sfe_ipv6 0
+}
+
+stop() {
+	[ -d /sys/module/shortcut_fe_cm ] && rmmod shortcut_fe_cm
+	[ -d /sys/module/fast_classifier ] && rmmod fast_classifier
+}
-- 
2.17.1

